### PR TITLE
KAA-1392: [C++ SDK] QNX 6.5.0 support.

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
@rasendubi and @GamovCoder please review.

File: client/client-multi/client-cpp/CMakeLists.txt [hunk 1]
```diff
@@ -221,7 +221,7 @@ endif()
 if(CMAKE_COMPILER_IS_GNUCXX)
     list(APPEND KAA_COMPILE_OPTIONS 
                     -Wall
-                    -std=c++11)
```
Comment (rasendubi) [inline client/client-multi/client-cpp/CMakeLists.txt:224]: We should stick with the standard as much as possible. GNU extensions are banned unless we can't leave without them. It's ok to enable them for QNX only, though.

File: client/client-multi/client-cpp/docs/qnx/avro-cpp-disable-tests.patch [hunk 2]
```diff
@@ -0,0 +1,92 @@
+diff -crB avro-cpp-1.7.5/CMakeLists.txt avro-cpp-1.7.5_patched/CMakeLists.txt
```
Comment (rasendubi) [inline client/client-multi/client-cpp/docs/qnx/avro-cpp-disable-tests.patch]: I believe we already have this patch in the source tree.

File: client/client-multi/client-cpp/toolchains/qnx-x86.cmake [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+set(CMAKE_SYSTEM_VERSION 6.5.0)
+
+set(CMAKE_C_COMPILER $ENV{QNX_HOST}/usr/bin/i486-pc-nto-qnx${CMAKE_SYSTEM_VERSION}-gcc)
+set(CMAKE_CXX_COMPILER $ENV{QNX_HOST}/usr/bin/i486-pc-nto-qnx${CMAKE_SYSTEM_VERSION}-g++) 
+
+set(CMAKE_LIBRARY_ARCHITECTURE x86)
+
+set(CMAKE_FIND_ROOT_PATH
+        ${CMAKE_FIND_ROOT_PATH}
+        $ENV{QNX_TARGET}
```
Comment (rasendubi) [inline client/client-multi/client-cpp/toolchains/qnx-x86.cmake]: We prefer cmake variables over environment ones.

File: client/client-multi/client-cpp/toolchains/qnx-x86.cmake [hunk 4]
```diff
@@ -0,0 +1,19 @@
+set(CMAKE_SYSTEM_NAME QNX)
+set(CMAKE_SYSTEM_VERSION 6.5.0)
+
+set(CMAKE_C_COMPILER $ENV{QNX_HOST}/usr/bin/i486-pc-nto-qnx${CMAKE_SYSTEM_VERSION}-gcc)
```
Comment (rasendubi) [inline client/client-multi/client-cpp/toolchains/qnx-x86.cmake]: We prefer cmake variables over environment ones.

File: client/client-multi/client-cpp/CMakeLists.txt [hunk 5]
```diff
@@ -448,10 +448,10 @@ if(WIN32 AND "${CMAKE_BUILD_TYPE}" STREQUAL "Debug")
             dbghelp)
 endif()
 
-if(NOT WIN32)
+if(CMAKE_SYSTEM_NAME STREQUAL "QNX")
     set(KAA_THIRDPARTY_LIBRARIES
             ${KAA_THIRDPARTY_LIBRARIES}
-            pthread)
+            socket)
```
Comment (rasendubi) [inline client/client-multi/client-cpp/CMakeLists.txt:461]: pthread was removed for all non-windows platforms. We should recheck if it's really needed.

File: client/client-multi/client-cpp/docs/qnx/avro-cpp-disable-tests.patch (hunk 2 above)
Comment (ghost) [inline client/client-multi/client-cpp/docs/qnx/avro-cpp-disable-tests.patch]: Great. Could you give me a link? So I can use it in building instructions.

Comment (ohamov) [reply client/client-multi/client-cpp/docs/qnx/avro-cpp-disable-tests.patch]: https://raw.githubusercontent.com/kaaproject/kaa/967970ec57fb0a62c23ffe573385bf0d0299d977/client/client-multi/client-cpp/tools/avro-cpp-disable-tests.patch

File: client/client-multi/client-cpp/CMakeLists.txt (hunk 1 above)
Comment (ghost) [reply client/client-multi/client-cpp/CMakeLists.txt:224]: Ok. Will enable only for QNX

File: client/client-multi/client-cpp/toolchains/qnx-x86.cmake [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        ${CMAKE_FIND_ROOT_PATH}
+        $ENV{QNX_TARGET}
+        $ENV{QNX_TARGET}/${CMAKE_LIBRARY_ARCHITECTURE})
+
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
+set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)
+
+add_definitions(-DQNX_650_CPP11_TO_STRING_PATCH)
```
Comment (GamovCoder) [inline client/client-multi/client-cpp/toolchains/qnx-x86.cmake]: add empty line

File: client/client-multi/client-cpp/CMakeLists.txt (hunk 5 above)
Comment (ghost) [reply client/client-multi/client-cpp/CMakeLists.txt:461]: Will use ${CMAKE_THREAD_LIBS_INIT} instead.

File: client/client-multi/client-cpp/docs/qnx/avro-cpp-disable-tests.patch (hunk 2 above)
Comment (ghost) [reply client/client-multi/client-cpp/docs/qnx/avro-cpp-disable-tests.patch]: Ok. Will re-use the patch from tools/

**Comment by rasendubi** [comment]
@kaa-leeroy retest this please

**Pull request closed without merging**